File: apps/android/app/src/main/java/dev/flextranslate/foundation/M2m100MtProvider.kt
package dev.flextranslate.foundation

import android.util.Log
import java.io.File

/**
 * Настоящий on-device машинный перевод через M2M-100 ([M2m100OnnxEngine]), реализует
 * [TranslationProvider].
 *
 * G005/WS4 (A2): каждый ненулевой [TranslationResult.text] — ПОДЛИННЫЙ вывод модели. Никаких
 * выдуманных/заготовленных переводов нет: если файлов модели нет или перевод упал, провайдер
 * честно вернёт [TranslationResult.unsupportedReason] (гейт), но не фейковый текст.
 *
 * Движок создаётся лениво при первом обращении и переиспользуется (грузить две ONNX-сессии дорого).
 * Направление берём из строки [languagePair] вида `"src->tgt"`, например `"en->ru"`. M2M-100 тянет
 * RU↔EN и RU↔ZH напрямую через форсированный target-токен — без пивота через английский.
 *
 * Заявления в support-matrix всё ещё требуют бенчмарков WS6; рабочая демка A2 — не заявка на релиз.
 * [deviceTier] принимаем ради единообразия интерфейса и телеметрии, на поведение он не влияет.
 */
class M2m100MtProvider(
    private val spec: MtModelSpec.Seq2SeqOnnx,
    private val modelDir: File,
) : TranslationProvider {

    override val providerId: String = "m2m100:${spec.modelId}"

    private sealed interface State {
        data object Uninitialized : State
        data class Ready(val engine: M2m100OnnxEngine) : State
        data object MissingModel : State
        data class Failed(val reason: String) : State
    }

    private var state: State = State.Uninitialized

    override fun translate(text: String, languagePair: String, deviceTier: String): TranslationResult {
        val trimmed = text.trim()
        if (trimmed.isEmpty()) return TranslationResult(text = null, unsupportedReason = null)

        val direction = MtDirection.parse(languagePair)
            ?: return TranslationResult(
                text = null,
                unsupportedReason = "пара $languagePair не поддерживается этой моделью",
            )

        return when (val ready = ensureEngine()) {
            is State.Ready -> runTranslate(ready.engine, trimmed, direction)
            State.MissingModel -> TranslationResult(
                text = null,
                unsupportedReason = "MT-модель ${spec.modelId} не установлена",
            )
            is State.Failed -> TranslationResult(
                text = null,
                unsupportedReason = "MT-движок недоступен: ${ready.reason}",
            )
            State.Uninitialized -> TranslationResult(
                text = null,
                unsupportedReason = "MT-движок не инициализирован",
            )
        }
    }

    private fun runTranslate(engine: M2m100OnnxEngine, text: String, direction: MtDirection): TranslationResult =
        runCatching {
            val output = engine.translate(text, direction.source, direction.target)
            TranslationResult(text = output, unsupportedReason = null)
        }.getOrElse { t ->
            Log.e(TAG, "translate failed ${direction.source}->${direction.target}", t)
            TranslationResult(text = null, unsupportedReason = "ошибка перевода: ${t.message ?: t.javaClass.simpleName}")
        }

    /** Честная готовность без полной загрузки сессии, когда модели попросту нет. */
    fun isModelInstalled(): Boolean = spec.isInstalled(modelDir)

    private fun ensureEngine(): State {
        val current = state
        if (current !is State.Uninitialized) return current
        if (!spec.isInstalled(modelDir)) {
            state = State.MissingModel
            return state
        }
        val next = M2m100OnnxEngine.create(spec, modelDir)
            ?.let { State.Ready(it) }
            ?: State.Failed("session init returned null")
        state = next
        return next
    }

    override fun close() {
        (state as? State.Ready)?.engine?.close()
        state = State.Uninitialized
    }

    private companion object {
        const val TAG = "M2m100MtProvider"
    }
}

/** Распарсенное направление перевода `"src->tgt"`, ограниченное демо-кодами языков M2M-100. */
data class MtDirection(val source: String, val target: String) {
    companion object {
        private val SUPPORTED = setOf("ru", "en", "zh")

        /** Парсит `"en->ru"`; null, если строка кривая или выходит за демо-набор RU/EN/ZH. */
        fun parse(languagePair: String): MtDirection? {
            val parts = languagePair.split("->", "-", "_", limit = 2)
            if (parts.size != 2) return null
            val source = parts[0].trim().lowercase()
            val target = parts[1].trim().lowercase()
            if (source !in SUPPORTED || target !in SUPPORTED || source == target) return null
            return MtDirection(source, target)
        }
    }
}

File: apps/ios/FlexTranslate/Sources/FlexTranslate/M2m100MtProvider.swift
import Foundation

// Real on-device machine translation via M2M-100 (M2m100OnnxEngine) conforming to
// TranslationProvider.
//
// A2 discipline: every non-nil TranslationResult.text is GENUINE model output.
// No fabricated/canned translation anywhere. If model files are absent or a
// translation fails, the provider returns an honest unsupportedReason, never fake text.
//
// The engine is created lazily on first use and reused across calls (loading three
// ONNX sessions is expensive). Direction comes from a "src->tgt" languagePair string,
// e.g. "ru->en". M2M-100 handles RU↔EN and RU↔ZH via a forced target-language token
// — no English pivot required.
//
// Not @MainActor: translate() may be called from any thread; lazy engine init is
// protected by a lock. State transitions are monotone (Uninitialized → one terminal).
final class M2m100MtProvider: TranslationProvider {
    let providerId: String

    private let spec: MtModelSpec
    private let modelDir: URL

    private enum State {
        case uninitialized
        case ready(M2m100OnnxEngine)
        case missingModel
        case failed(String)
    }

    private let stateLock = NSLock()
    private var _state: State = .uninitialized

    init(spec: MtModelSpec, modelDir: URL) {
        self.spec = spec
        self.modelDir = modelDir
        self.providerId = "m2m100:\(spec.modelId)"
    }

    func translate(text: String, languagePair: String, deviceTier: String) -> TranslationResult {
        let trimmed = text.trimmingCharacters(in: .whitespaces)
        if trimmed.isEmpty {
            return TranslationResult(text: nil, unsupportedReason: nil)
        }

        guard let direction = MtDirection.parse(languagePair) else {
            return TranslationResult(
                text: nil,
                unsupportedReason: "пара \(languagePair) не поддерживается этой моделью"
            )
        }

        switch ensureEngine() {
        case .ready(let engine):
            return runTranslate(engine: engine, text: trimmed, direction: direction)
        case .missingModel:
            return TranslationResult(
                text: nil,
                unsupportedReason: "MT-модель \(spec.modelId) не установлена"
            )
        case .failed(let reason):
            return TranslationResult(
                text: nil,
                unsupportedReason: "MT-движок недоступен: \(reason)"
            )
        case .uninitialized:
            return TranslationResult(
                text: nil,
                unsupportedReason: "MT-движок не инициализирован"
            )
        }
    }

    var isModelInstalled: Bool {
        MtModelStore.shared.isInstalled(spec)
    }

    func close() {
        stateLock.lock()
        let current = _state
        _state = .uninitialized
        stateLock.unlock()
        if case .ready(let engine) = current {
            engine.close()
        }
    }

    // MARK: - Private

    private func runTranslate(
        engine: M2m100OnnxEngine,
        text: String,
        direction: MtDirection
    ) -> TranslationResult {
        guard let output = engine.translate(
            text: text,
            sourceLang: direction.source,
            targetLang: direction.target
        ) else {
            return TranslationResult(
                text: nil,
                unsupportedReason: "ошибка перевода: движок вернул nil"
            )
        }
        return TranslationResult(text: output, unsupportedReason: nil)
    }

    private func ensureEngine() -> State {
        stateLock.lock()
        let current = _state
        stateLock.unlock()

        if case .uninitialized = current {} else { return current }

        if !MtModelStore.shared.isInstalled(spec) {
            stateLock.lock()
            _state = .missingModel
            stateLock.unlock()
            return .missingModel
        }

        let dir = MtModelStore.shared.modelDir(for: spec)
        if let engine = M2m100OnnxEngine.create(spec: spec, modelDir: dir) {
            let next = State.ready(engine)
            stateLock.lock()
            _state = next
            stateLock.unlock()
            return next
        } else {
            let next = State.failed("session init returned nil")
            stateLock.lock()
            _state = next
            stateLock.unlock()
            return next
        }
    }
}

// MARK: - Direction parsing

// A parsed "src->tgt" translation direction restricted to M2M-100 demo language codes.
struct MtDirection: Equatable {
    let source: String
    let target: String

    private static let supported: Set<String> = ["ru", "en", "zh"]

    // Parse "en->ru"; nil when malformed or outside the RU/EN/ZH demo scope.
    static func parse(_ languagePair: String) -> MtDirection? {
        // Accept "->" or "-" or "_" as separator; take at most 2 parts.
        let separators: [String] = ["->", "-", "_"]
        var parts: [String]?
        for sep in separators {
            let split = languagePair.components(separatedBy: sep)
            if split.count == 2 {
                parts = split
                break
            }
        }
        guard let parts else { return nil }
        let source = parts[0].trimmingCharacters(in: .whitespaces).lowercased()
        let target = parts[1].trimmingCharacters(in: .whitespaces).lowercased()
        guard supported.contains(source), supported.contains(target), source != target else {
            return nil
        }
        return MtDirection(source: source, target: target)
    }
}
